### PR TITLE
Improve/integrationtest psen scan upd interface

### PR DESCRIPTION
**Pull request by hslusarek**
- The PR tries to remedy issue #23.
- The PR also tries to improve the primitive communication interface original named `UDPInteface` which is now named `ScannerCommunicationInterface`.

### CODE REVIEW

File: test/unit_tests/unittest_scanner_communication_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+#include <psen_scan/scanner_communication_interface.h>
+#include <psen_scan/psen_scan_udp_interface.h>
+#include <psen_scan/udp_read_timeout_exception.h>
+
+using namespace psen_scan;
+
+TEST(ScannerCommunicationIntefaceTests, testFailingWriteOperation)
+{
+  const uint32_t udp_port{ 55000 };
```
Comment (SansoneG) [inline test/unit_tests/unittest_scanner_communication_interface.cpp]: Could avoid duplicate code by using a test fixture.

File: test/unit_tests/unittest_scanner_communication_interface.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  const std::string ip_addr{ "127.0.0.1" };
+
+  boost::asio::io_service io_service;
+  std::unique_ptr<ScannerCommunicationInterface> comm_interface =
+      std::unique_ptr<PSENscanUDPInterface>(new PSENscanUDPInterface(io_service, ip_addr, udp_port));
+
+  boost::array<char, 10> write_buf = { "Hello!" };
+  ASSERT_THROW(comm_interface->write(boost::asio::buffer(write_buf)), ScannerWriteFailed);
```
Comment (SansoneG) [inline test/unit_tests/unittest_scanner_communication_interface.cpp:38]: Why will the write operation fail?

**Comment by hslusarek** [comment]
@SansoneG Ping: Ready for review. What do you think?

File: src/psen_scan_udp_interface.cpp [hunk 3]
```diff
@@ -32,18 +32,49 @@ static const uint64_t TIMEOUT_LOOP_SLEEP_DURATION_MS = 5;
...
-  socket_write_.connect(udp_write_endpoint_);
-  socket_read_.connect(udp_read_endpoint_);
 }
 
+PSENscanUDPInterface::~PSENscanUDPInterface()
+{
+  close();
+}
+
+// LCOV_EXCL_START
```
Comment (SansoneG) [inline src/psen_scan_udp_interface.cpp]: Why are these excluded?

File: test/integration_tests/integrationtest_psen_scan_udp_interface.cpp [hunk 4]
```diff
@@ -13,47 +13,86 @@
...
+public:
+  void SetUp() override;
+
+protected:
+  std::future<std::size_t> startAsyncReadOperation(boost::asio::mutable_buffers_1& read_buf);
 
-  MockUDPServer mock_udp_server(io_service_1);
+protected:
+  MockUDPServer mock_udp_server_;
+  std::unique_ptr<ScannerCommunicationInterface> scanner_interface_{ new PSENscanUDPInterface(IP_ADDRESS, UDP_PORT) };
```
Comment (SansoneG) [inline test/integration_tests/integrationtest_psen_scan_udp_interface.cpp]: does this need to be a pointer?

File: test/unit_tests/unittest_scanner_communication_interface.cpp [hunk 5]
```diff
@@ -48,3 +51,36 @@ TEST(ScannerCommunicationIntefaceTests, testReadTimeout)
...
+TEST(ScannerCommunicationIntefaceTests, testScannerWriteFailedForCompleteCoverage)
+{
+  std::string except_str{ "DummyText" };
+  std::unique_ptr<ScannerWriteFailed> e(new ScannerWriteFailed(except_str));
+  EXPECT_EQ(except_str, e->what());
+}
+
+TEST(ScannerCommunicationIntefaceTests, testScannerScannerReadTimeoutForCompleteCoverage)
+{
+  std::unique_ptr<ScannerReadTimeout> e(new ScannerReadTimeout());
```
Comment (SansoneG) [inline test/unit_tests/unittest_scanner_communication_interface.cpp:69]: Add check that string is equal to "Timeout while waiting for message from scanner"?

**Review by SansoneG — APPROVED** [review]

**Pull request merged**